cmd/snapctl/task.go

```diff
@@ -114,14 +114,226 @@ func stringValToInt(val string) (int, error) {
+}
+
+// Parses the command-line parameters and uses them to override the underlying schedule
+// in this task (if any) or set a schedule for that task (if one is not already defined)
+func (t *task) setWindowedSchedule(start *time.Time, stop *time.Time, duration *time.Duration) error {
+	// if this task does not have an existing schedule, create an empty
+	// windowed schedule
+	if t.Schedule == nil {
+		t.Schedule = &client.Schedule{
+			Type: "windowed",
+		}
+	} else if t.Schedule.Type == "" {
+		// else if there is an empty schedule already defined for this task,
+		// set the type for that schedule to 'windowed'
+		t.Schedule.Type = "windowed"
+	} else if t.Schedule.Type != "windowed" {
+		// else if the task's existing schedule is not a 'windowed' schedule,
+		// then return an error
+		return fmt.Errorf("Usage error (schedule type mismatch); cannot replace existing schedule of type '%v' with a new, 'windowed' schedule", t.Schedule.Type)
+	}
+	// grab the task schedule's start/stop time (will use these later to fill in a
+	// missing boundary value, if we find one missing in our input arguments)
+	scheduleStart := t.Schedule.StartTime
+	scheduleStop := t.Schedule.StopTime
+	// if a duration was passed in, determine the start and stop times for our new
+	// 'windowed' schedule from the input parameters
+	if duration != nil {
+		// if start and stop were both defined, then return an error (since all three parameters cannot be used
+		// to define the 'windowed' schedule)
+		if start != nil && stop != nil {
+			return fmt.Errorf("Usage error (too many parameters); only two of the parameters that define the window (start time, stop time and duration) can be specified for a 'windowed' schedule")
+		}
+		// if start is set and stop is not then use duration to create stop
+		if start != nil && stop == nil {
+			newStop := start.Add(*duration)
+			stop = &newStop
+		} else if stop != nil && start == nil {
+			// else if stop is set and start is not then use duration to create start
+			newStart := stop.Add(*duration * -1)
+			start = &newStart
+		} else {
+			// else, the start and stop are both undefined but a duration was passed in,
+			// so use the current date/time (plus the 'createTaskNowPad' value) as the
+			// start date/time and construct a stop date/time from that start date/time
+			// and the duration
+			newStart := time.Now().Add(createTaskNowPad)
+			newStop := newStart.Add(*duration)
+			start = &newStart
+			stop = &newStop
+		}
+		// check to make sure that our new time window is a valid schedule
+		if err := t.checkScheduleWindow(*start, *stop); err != nil {
+			return err
+		}
+		// if it's a valid window, then use the (new) start/stop values to set the
+		// start/stop date/times for our new schedule
+		t.Schedule.StartTime = start
+		t.Schedule.StopTime = stop
+		// and return a nil error (indicating success)
+		return nil
+	} else if (start == nil && scheduleStart == nil) || (stop == nil && scheduleStop == nil) {
```

Comment:
With the first `if` returning if duration is not nil, this does not need to be an `else if`.